icechunk/src/session.rs

```diff
@@ -0,0 +1,1975 @@
+        assert!(ds.get_group(&"/b".try_into().unwrap()).await.is_err());
+        assert!(ds.get_group(&"/b/bb".try_into().unwrap()).await.is_err());
+        Ok(())
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_all_chunks_iterator() -> Result<(), Box<dyn Error>> {
+        let repository = create_memory_store_repository().await;
+        let mut ds = repository.writeable_session("main").await?;
+
+        // add a new array and retrieve its node
+        ds.add_group(Path::root()).await?;
+        let zarr_meta = ZarrArrayMetadata {
+            shape: vec![1, 1, 2],
+            data_type: DataType::Int32,
+            chunk_shape: ChunkShape(vec![NonZeroU64::new(2).unwrap()]),
+            chunk_key_encoding: ChunkKeyEncoding::Slash,
+            fill_value: FillValue::Int32(0),
+            codecs: vec![Codec { name: "mycodec".to_string(), configuration: None }],
+            storage_transformers: Some(vec![StorageTransformer {
+                name: "mytransformer".to_string(),
+                configuration: None,
+            }]),
+            dimension_names: Some(vec![Some("t".to_string())]),
+        };
+
+        let new_array_path: Path = "/array".try_into().unwrap();
+        ds.add_array(new_array_path.clone(), zarr_meta.clone()).await?;
+        // we 3 chunks
+        ds.set_chunk_ref(
+            new_array_path.clone(),
+            ChunkIndices(vec![0, 0, 0]),
+            Some(ChunkPayload::Inline("hello".into())),
+        )
+        .await?;
+        ds.set_chunk_ref(
+            new_array_path.clone(),
+            ChunkIndices(vec![0, 0, 1]),
+            Some(ChunkPayload::Inline("hello".into())),
+        )
+        .await?;
+        ds.set_chunk_ref(
+            new_array_path.clone(),
+            ChunkIndices(vec![1, 0, 0]),
+            Some(ChunkPayload::Inline("hello".into())),
+        )
+        .await?;
+        let snapshot_id = ds.commit("commit", None).await?;
+
+        let ds =
+            repository.readonly_session(&VersionInfo::SnapshotId(snapshot_id)).await?;
+        let coords = ds
+            .all_chunks()
+            .await?
+            .map_ok(|(_, chunk)| chunk.coord)
+            .try_collect::<HashSet<_>>()
+            .await?;
+        assert_eq!(
+            coords,
+            vec![
+                ChunkIndices(vec![0, 0, 0]),
+                ChunkIndices(vec![0, 0, 1]),
+                ChunkIndices(vec![1, 0, 0])
+            ]
+            .into_iter()
+            .collect()
+        );
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_manifests_shrink() -> Result<(), Box<dyn Error>> {
+        let in_mem_storage =
+            Arc::new(ObjectStorage::new_in_memory_store(Some("prefix".into()))?);
+        let storage: Arc<dyn Storage + Send + Sync> = in_mem_storage.clone();
+        let repository =
+            Repository::create(RepositoryConfig::default(), storage, None).await?;
```

Comment:
This shwos teh session usage. @paraseba is this the workflow you expected? with having to get a new session after each commit?